x-pack/filebeat/input/s3/s3_integration_test.go

```diff
@@ -0,0 +1,417 @@
+			MessageAttributeNames: []string{"All"},
+			MaxNumberOfMessages:   &maxNumberOfMessage,
+			VisibilityTimeout:     &visibilityTimeout,
+			WaitTimeSeconds:       &waitTimeSecond,
+		})
+
+	output, err := req.Send(ctx)
+	if err != nil {
+		t.Fatalf("failed to receive message from SQS: %v", output)
+	}
+
+	var oldMessageHandles []string
+	for _, message := range output.Messages {
+		oldMessageHandles = append(oldMessageHandles, *message.ReceiptHandle)
+	}
+
+	return oldMessageHandles
+}
+
+func (input *s3Input) deleteAllMessages(t *testing.T, awsConfig awssdk.Config, queueURL string, visibilityTimeout int64, svcSQS sqsiface.ClientAPI) error {
+	var messageReceiptHandles []string
+	init := true
+	for init || len(messageReceiptHandles) > 0 {
+		init = false
+		messageReceiptHandles = collectOldMessages(t, queueURL, visibilityTimeout, svcSQS)
+		for _, receiptHandle := range messageReceiptHandles {
+			err := input.deleteMessage(queueURL, receiptHandle, svcSQS)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func defaultTestConfig() *common.Config {
+	return common.MustNewConfigFrom(map[string]interface{}{
+		"queue_url": os.Getenv("QUEUE_URL"),
+	})
+}
+
+func runTest(t *testing.T, cfg *common.Config, run func(input *s3Input, out *stubOutleter, t *testing.T)) {
+	clientCancel := testSetup(t)
+	defer clientCancel()
+
+	// Simulate input.Context from Filebeat input runner.
+	inputCtx := newInputContext()
+	defer close(inputCtx.Done)
+
+	// Stub outlet for receiving events generated by the input.
+	eventOutlet := newStubOutlet()
+	defer eventOutlet.Close()
+
+	connector := channel.ConnectorFunc(func(_ *common.Config, _ beat.ClientConfig) (channel.Outleter, error) {
+		return eventOutlet, nil
+	})
+
+	in, err := NewInput(cfg, connector, inputCtx)
+	if err != nil {
+		t.Skipf("Skipping: %v", err)
+	}
+	s3Input := in.(*s3Input)
+	defer s3Input.Stop()
+
+	run(s3Input, eventOutlet, t)
+}
+
+func newInputContext() input.Context {
+	return input.Context{
+		Done: make(chan struct{}),
+	}
+}
+
+type stubOutleter struct {
+	sync.Mutex
+	cond   *sync.Cond
+	done   bool
+	Events []beat.Event
+}
+
+func newStubOutlet() *stubOutleter {
+	o := &stubOutleter{}
+	o.cond = sync.NewCond(o)
+	return o
+}
+
+func (o *stubOutleter) waitForEvents(numEvents int) ([]beat.Event, bool) {
+	o.Lock()
+	defer o.Unlock()
+
+	for len(o.Events) < numEvents && !o.done {
+		o.cond.Wait()
+	}
+
+	size := numEvents
+	if size >= len(o.Events) {
+		size = len(o.Events)
+	}
+
+	out := make([]beat.Event, size)
+	copy(out, o.Events)
+	return out, len(out) == numEvents
+}
+
+func (o *stubOutleter) Close() error {
+	o.Lock()
+	defer o.Unlock()
+	o.done = true
+	return nil
+}
+
+func (o *stubOutleter) Done() <-chan struct{} { return nil }
+
+func (o *stubOutleter) OnEvent(event beat.Event) bool {
+	o.Lock()
+	defer o.Unlock()
+	o.Events = append(o.Events, event)
+	o.cond.Broadcast()
+	return !o.done
+}
+
+func TestS3Input(t *testing.T) {
```

Comment:
Use the goroutine checker also here? :slightly_smiling_face:

Reply:
I tried and it failed and it's not because of the filebeat s3 input code (I don't think). I tried debugging and seems like it's because this is testing with real aws account and calling real aws api. For some reason, some aws apis using aws-sdk-go-v2 is having problems with not closing goroutines hmm

Reply:
This is interesting, because we may have leaks because of that. Lets keep an eye on it.